duopull/duopull.go

```diff
@@ -0,0 +1,698 @@
+	return "Basic " + base64.StdEncoding.EncodeToString([]byte(auth)), ds
+}
+
+// logRecords represents a response for a log request from the Duo API.
+//
+// The actual event information is treated as arbitrary JSON. The stat field is
+// specifically included so we can inspect the value to confirm the request was
+// successful.
+//
+// See also https://duo.com/docs/adminapi#api-details
+type logRecords struct {
+	Stat     string        `json:"stat"`
+	Response []interface{} `json:"response"`
+}
+
+type logRecordsV2 struct {
+	Stat     string               `json:"stat"`
+	Response logRecordsV2Response `json:"response"`
+}
+
+type logRecordsV2Response struct {
+	Authlogs []interface{} `json:"authlogs"`
+}
+
+// emitEvent is an event which will be submitted to Kinesis
+//
+// Since the event itself contains no reference to the class of the event (e.g., authentication,
+// administrator, etc) the path used to request the log is included here so the event types
+// can be differentiated by the stream consumer.
+type emitEvent struct {
+	Path  string      `json:"path"`  // The request path (e.g., /api/v1/logs/telephony)
+	Event interface{} `json:"event"` // The actual event
+}
+
+func (e *emitEvent) toInterface() (map[string]interface{}, error) {
+	ret := make(map[string]interface{})
+	buf, err := json.Marshal(e)
+	if err != nil {
+		return ret, err
+	}
+	err = json.Unmarshal(buf, &ret)
+	return ret, err
+}
+
+// getTimestamp extracts the timestamp value from e as an integer
+func (e *emitEvent) getTimestamp() (int, error) {
+	// Define a pseudo-struct for extraction of the timestamp instead of using
+	// type assertions and dealing with float64 conversion
+	type pse struct {
+		Timestamp int `json:"timestamp"`
+	}
+	var p pse
+	buf, err := json.Marshal(e.Event)
+	if err != nil {
+		return 0, err
+	}
+	err = json.Unmarshal(buf, &p)
+	if err != nil {
+		return 0, err
+	}
+	if p.Timestamp == 0 {
+		return 0, fmt.Errorf("event had no timestamp")
+	}
+	return p.Timestamp, nil
+}
+
+// emitter stores all collected events for distribution to Kinesis
+type emitter struct {
+	events []emitEvent
+}
+
+// emit batches collected events to the configured Kinesis stream
+func (e *emitter) emit() error {
+	if debug == debugDuo { // Duo debug, just write events to stdout
+		for _, v := range e.events {
+			cv, err := v.toInterface()
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			out, err := toMozLog(cv)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			buf, err := json.Marshal(out)
+			if err != nil {
+				return err
+			}
+			log.Infof("%v\n", string(buf))
+		}
+		return nil
+	}
+
+	logger := cfg.stackdriverClient.Logger(LOGGER_NAME)
+
+	for _, v := range e.events {
+		cv, err := v.toInterface()
+		if err != nil {
+			return err
+		}
+		out, err := toMozLog(cv)
+		if err != nil {
+			return err
+		}
+		e, err := json.Marshal(out)
+		if err != nil {
+			return err
+		}
+		logger.Log(stackdriver.Entry{Payload: json.RawMessage(e)})
```

Comment:
Unrelated to this PR, but we may need to adjust the pipeline parser for this (just adding it here as a note).